File: frontend/packages/arch/api-schema/src/idl/corporation/department.ts
/*
 * Copyright 2025 coze-plus Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as base from './../base';
export { base };
import * as common from './common';
export { common };
export interface DepartmentData {
  id: string,
  corp_id: string,
  parent_id?: string,
  name: string,
  code?: string,
  level: number,
  full_path?: string,
  leader_id?: string,
  sort: number,
  status: common.DepartmentStatus,
  out_department_id?: string,
  department_source?: common.DataSource,
  creator_id: string,
  created_at: string,
  updated_at: string,
  deleted_at?: string,
}
export interface CreateDepartmentRequest {
  name: string,
  corp_id: string,
  parent_id?: string,
  code?: string,
  leader_id?: string,
  sort?: number,
  out_department_id?: string,
  department_source?: common.DataSource,
}
export interface CreateDepartmentResponse {
  code: number,
  msg: string,
  data?: DepartmentData,
}
export interface UpdateDepartmentRequest {
  id: string,
  name?: string,
  parent_id?: string,
  code?: string,
  leader_id?: string,
  sort?: number,
  status?: common.DepartmentStatus,
  out_department_id?: string,
  department_source?: common.DataSource,
}
export interface UpdateDepartmentResponse {
  code: number,
  msg: string,
}
export interface DeleteDepartmentRequest {
  id: string
}
export interface DeleteDepartmentResponse {
  code: number,
  msg: string,
}
export interface GetDepartmentRequest {
  id: string
}
export interface GetDepartmentResponse {
  code: number,
  msg: string,
  data?: DepartmentData,
}
export interface ListDepartmentRequest {
  corp_id: string,
  parent_id?: string,
  keyword?: string,
  status?: common.DepartmentStatus,
  page?: number,
  page_size?: number,
}
export interface ListDepartmentResponse {
  code: number,
  msg: string,
  data: DepartmentData[],
  total: string,
}
export interface DepartmentId {
  id: string,
  sort: number,
}
export interface SortDepartmentRequest {
  parent_id: string,
  department_ids: DepartmentId[],
}
export interface SortDepartmentResponse {
  code: number,
  msg: string,
}
/** 部门树节点 */
export interface DepartmentTreeNode {
  id: string,
  corp_id: string,
  parent_id?: string,
  name: string,
  code?: string,
  level: number,
  full_path?: string,
  leader_id?: string,
  sort: number,
  status: common.DepartmentStatus,
  employee_count: number,
  children: DepartmentTreeNode[],
  has_children: boolean,
  is_expanded: boolean,
}
/** 获取部门树请求 */
export interface GetDepartmentTreeRequest {
  corp_id: string,
  parent_id?: string,
  /** 展开深度，0表示全部展开 */
  depth?: number,
  include_employee_count?: boolean,
  status?: common.DepartmentStatus,
}
/** 获取部门树响应 */
export interface GetDepartmentTreeResponse {
  code: number,
  msg: string,
  data: DepartmentTreeNode[],
}

File: backend/domain/corporation/entity/department.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entity

import (
	"errors"
	"strings"
)

// Department entity represents a department in the domain
type Department struct {
	ID           int64             `json:"id"`
	CorpID       int64             `json:"corp_id"`
	ParentDeptID *int64            `json:"parent_dept_id"`
	Name         string            `json:"name"`
	Code         *string           `json:"code"`
	Level        int32             `json:"level"`
	FullPath     string            `json:"full_path"`
	LeaderID     *int64            `json:"leader_id"`
	Sort         int32             `json:"sort"`
	Status       DepartmentStatus  `json:"status"`
	OutDeptID    *string           `json:"out_dept_id"`
	DeptSource   DepartmentSource  `json:"dept_source"`
	CreatorID    int64             `json:"creator_id"`
	CreatedAt    int64             `json:"created_at"`
	UpdatedAt    int64             `json:"updated_at"`
	DeletedAt    *int64            `json:"deleted_at"`
	
	// Aggregated fields (not stored in DB)
	Children  []*Department `json:"children,omitempty"`
	Employees []*Employee   `json:"employees,omitempty"`
}

// DepartmentSource defines data source enum
type DepartmentSource int32

const (
	DepartmentSourceManual       DepartmentSource = 0
	DepartmentSourceEnterpriseWX DepartmentSource = 1
	DepartmentSourceDingTalk     DepartmentSource = 2
	DepartmentSourceFeishu       DepartmentSource = 3
)

// DepartmentStatus defines department status enum
type DepartmentStatus int32

const (
	DepartmentStatusActive   DepartmentStatus = 1
	DepartmentStatusInactive DepartmentStatus = 2
)

// CreateDepartmentMeta contains metadata for creating department
type CreateDepartmentMeta struct {
	CorpID       int64            `json:"corp_id"`
	ParentDeptID *int64           `json:"parent_dept_id"`
	Name         string           `json:"name"`
	Sort         int32            `json:"sort"`
	OutDeptID    *string          `json:"out_dept_id"`
	DeptSource   DepartmentSource `json:"dept_source"`
	CreatorID    int64            `json:"creator_id"`
}

// UpdateDepartmentMeta contains metadata for updating department
type UpdateDepartmentMeta struct {
	ParentDeptID *int64           `json:"parent_dept_id"`
	Name         *string          `json:"name"`
	Sort         *int32           `json:"sort"`
	OutDeptID    *string          `json:"out_dept_id"`
	DeptSource   *DepartmentSource `json:"dept_source"`
}

// DepartmentListFilter contains filter criteria for listing departments
type DepartmentListFilter struct {
	CorpID       *int64           `json:"corp_id"`
	ParentDeptID *int64           `json:"parent_dept_id"`
	Status       *DepartmentStatus `json:"status"`
	DeptSource   *DepartmentSource `json:"dept_source"`
	CreatorID    *int64           `json:"creator_id"`
	Keyword      *string          `json:"keyword"` // Search in name
	Limit        int              `json:"limit"`
	Page         int              `json:"page"`
}

// Business methods

// ValidateName validates department name
func (d *Department) ValidateName(name string) error {
	name = strings.TrimSpace(name)
	if len(name) == 0 {
		return errors.New("department name cannot be empty")
	}
	if len(name) > 100 {
		return errors.New("department name cannot exceed 100 characters")
	}
	return nil
}

// CanDelete checks if department can be deleted
func (d *Department) CanDelete() bool {
	// Cannot delete if has children or employees
	return len(d.Children) == 0 && len(d.Employees) == 0
}

// IsRootDepartment checks if this is a root department
func (d *Department) IsRootDepartment() bool {
	return d.ParentDeptID == nil
}

// CanMoveTo checks if department can be moved to target parent
func (d *Department) CanMoveTo(targetParentID *int64) bool {
	// Cannot move to self or descendant
	if targetParentID != nil && *targetParentID == d.ID {
		return false
	}
	// Additional logic to prevent circular references would be needed
	return true
}

// AddChild adds a child department (for tree building)
func (d *Department) AddChild(child *Department) {
	if d.Children == nil {
		d.Children = make([]*Department, 0)
	}
	d.Children = append(d.Children, child)
}

// AddEmployee adds an employee (for aggregation)
func (d *Department) AddEmployee(emp *Employee) {
	if d.Employees == nil {
		d.Employees = make([]*Employee, 0)
	}
	d.Employees = append(d.Employees, emp)
}